In [1]:
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы glass<b>

In [19]:
glass = wordnet.synsets('glass')
for ss in glass:
    print(ss, ss.definition())

Synset('glass.n.01') a brittle transparent solid with irregular atomic structure
Synset('glass.n.02') a container for holding liquids while drinking
Synset('glass.n.03') the quantity a glass will hold
Synset('field_glass.n.01') a small refracting telescope
Synset('methamphetamine.n.01') an amphetamine derivative (trade name Methedrine) used in the form of a crystalline hydrochloride; used as a stimulant to the nervous system and as an appetite suppressant
Synset('looking_glass.n.01') a mirror; usually a ladies' dressing mirror
Synset('glass.n.07') glassware collectively
Synset('glass.v.01') furnish with glass
Synset('glass.v.02') scan (game in the forest) with binoculars
Synset('glass.v.03') enclose with glass
Synset('glass.v.04') put in a glass container
Synset('glaze.v.02') become glassy or take on a glass-like appearance


<b>2) Найти определение для лексемы glass в значении (а) "стекло" и в значении (b) "стакан"</b>

In [20]:
print(glass[0], glass[0].definition())
print(glass[1], glass[1].definition())

Synset('glass.n.01') a brittle transparent solid with irregular atomic structure
Synset('glass.n.02') a container for holding liquids while drinking


<b>3) Найдите два произвольных контекста для слова glass в значениях (a) "стекло" и (b) "стакан"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [25]:
sent1 = 'If you plan to install the painting opposite the window and it is near enough, the matte glass is better.'
sent2 = 'Now a glass of tea and a slice of bread a given for a breakfast.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['If', 'you', 'plan', 'to', 'install', 'the', 'painting', 'opposite', 'the', 'window', 'and', 'it', 'is', 'near', 'enough', 'the', 'matte', 'glass', 'is', 'better']
['Now', 'a', 'glass', 'of', 'tea', 'and', 'a', 'slice', 'of', 'bread', 'a', 'given', 'for', 'a', 'breakfast']


In [26]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'glass').definition())
print(lesk(sent2_tokens, 'glass').definition())

an amphetamine derivative (trade name Methedrine) used in the form of a crystalline hydrochloride; used as a stimulant to the nervous system and as an appetite suppressant
an amphetamine derivative (trade name Methedrine) used in the form of a crystalline hydrochloride; used as a stimulant to the nervous system and as an appetite suppressant


Алгоритм Леска ошибся в обоих случаях, но хотя бы определил часть речи.

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [27]:
for ss in glass[0].hypernyms():
    print(ss, ss.definition())
for ss in glass[1].hypernyms():
    print(ss, ss.definition())

Synset('solid.n.01') matter that is solid at room temperature and pressure
Synset('container.n.01') any object that can be used to hold things (especially a large metal boxlike object of standardized dimensions that can be loaded from one form of transport to another)


<b>5) Вычислите наименьшее расстояние между значением glass "стекло" и значениями лексемы material, а также glass "стакан" и значениями лексемы dishes.
Найти min (d(glass: "стекло", material), d(glass: "стекло", dishes)), а также min (d(glass: "стакан", material), d(glass: "стакан", dishes))</b>

In [29]:
material = wordnet.synsets('material')
dishes = wordnet.synsets('dishes')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(glass: "стекло", material)
dist1 = get_dist_sim(glass[0], material)[0]
print('min d(glass: "стекло", material): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(material[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(glass: "стекло", dishes)
dist2 = get_dist_sim(glass[0], dishes)[0]
print('min d(glass: "стекло", dishes): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(dishes[dist2.index(min(dist2))].definition()))

# min d(glass: "стакан", material)
dist3 = get_dist_sim(glass[1], material)[0]
print('min d(glass: "стакан", material): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(material[dist3.index(min(dist3))].definition()))

# min d(glass: "стакан", dishes)
dist4 = get_dist_sim(glass[1], dishes)[0]
print('min d(glass: "стакан", dishes): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(dishes[dist4.index(min(dist4))].definition()))

print('min (d(glass: "стекло", material), d(glass: "стекло", dishes)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(glass: "стакан", material), d(glass: "стакан", dishes)): {}'.format(min(min(dist3), min(dist4))))

min d(glass: "стекло", material): 4
closest lemma definition: the tangible substance that goes into the makeup of a physical object

min d(glass: "стекло", dishes): 6
closest lemma definition: a particular item of prepared food

min d(glass: "стакан", material): 4
closest lemma definition: artifact made by weaving or felting or knitting or crocheting natural or synthetic fibers

min d(glass: "стакан", dishes): 2
closest lemma definition: a piece of dishware normally used as a container for holding or serving food

min (d(glass: "стекло", material), d(glass: "стекло", dishes)): 4
min (d(glass: "стакан", material), d(glass: "стакан", dishes)): 2


Кратчайшее расстояние от стекла до material - 4, до dishes - 6.
Кратчайшее расстояние от стакана до material - 4, до dishes - 2. Выглядит хорошо.

<b>6) Вычислить двумя разными способами расстояние:
d(glass: "стакан", cup) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

In [32]:
cup = wordnet.synsets("cup")
for ss in cup:
    print(ss, ss.definition())

Synset('cup.n.01') a small open container usually used for drinking; usually has a handle
Synset('cup.n.02') the quantity a cup will hold
Synset('cup.n.03') any cup-shaped concavity
Synset('cup.n.04') a United States liquid unit equal to 8 fluid ounces
Synset('cup.n.05') cup-shaped plant organ
Synset('cup.n.06') a punch served in a pitcher instead of a punch bowl
Synset('cup.n.07') the hole (or metal container in the hole) on a golf green
Synset('cup.n.08') a large metal vessel with two handles that is awarded as a trophy to the winner of a competition
Synset('cup.v.01') form into the shape of a cup
Synset('cup.v.02') put into a cup
Synset('cup.v.03') treat by applying evacuated cups to the patient's skin


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [34]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(glass[1].path_similarity(cup[0]))

#Leacock-Chodorow Similarity
print(glass[1].lch_similarity(cup[0]))

#Wu-Palmer Similarity
print(glass[1].wup_similarity(cup[0]))
    
#Resnik Similarity
print(glass[1].res_similarity(cup[0], ic))
    
#Jiang-Conrath Similarity
print(glass[1].jcn_similarity(cup[0], ic))
    
#Lin Similarity
print(glass[1].lin_similarity(cup[0], ic))


0.3333333333333333
2.538973871058276
0.875
5.205612601196791
0.10673543596263503
0.5263462017356473


In [35]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [36]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [37]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>стакан/cup</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.3333333333</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>2.538973871058276</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.875</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>5.205612601196791</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>0.10673543596263503</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>0.5263462017356473</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>

Интуитивно кажется, что стакан/cup должен обладать наиболее высокими показателями, чем другие пары. А на втором месте за ним должен идти a living thing / an assemblage.

Значения первых трех индексов для стакан/cup и a living thing / an assemblage одинаковы. Значение Resnik Similarity для стакан/cup намного больше, чем у других пар, в том числе и a living thing / an assemblage. А значения последних двух индексов, наоборот, для стакан/cup меньше, чем для a living thing / an assemblage, и мне кажется, это странно.

Получается, первые 4 метрики лучше отражают интуитивное представление о близости понятий.